[dtensor] remove torchgen function schema and parse manually

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90241
* __->__ #90106
* #89970
* #89969
* #89968
* #89967

This PR get rids of torchgen FunctionSchema parsing and parse
it manually, it should resolve torchgen package issue and also
provide some perf wins when running DTensor eagerly

wanchaol committed:
[dtensor] remove torchgen function schema and parse manually

This PR get rids of torchgen FunctionSchema parsing and parse
it manually, it should resolve torchgen package issue and also
provide some perf wins when running DTensor eagerly

[ghstack-poisoned]

awgu commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -68,20 +66,20 @@ class OpSchema(object):
-        self.is_out_variant = (
-            schema_kind == SchemaKind.out  # pyre-ignore [16] pyre bad at enum
-        )
+        # simple analysis of function schema to determine
```
Conceptual: Does this PR regress the computation of `self.is_inplace` and `self._is_out_variant` in some cases because you are using a heuristic instead of `SchemaKind`?

I am curious how you think about the tradeoff of getting rid of the `torchgen` dependency versus possibly having incorrect values for `is_inplace` and `is_out_variant` in some cases (if at all).

wanchaol replied:
Good point, from the current operator set registered, i think it should work fine as all of the op signature naming following the same convention for inplace and out variant (`_` and `.out`), so I think this heuristic is relatively safe to use and should not regress existing registered operators. However, it does rely on operators following the same convention of naming, which I think it might get changed. Given that ATen op set is relatively stable at this point, I think we can go with this approach as it's simple and very fast without parsing FunctionSchema, if ever op signature gets changed, the heuristic need to be adapted to accommodate.

awgu commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -288,10 +284,13 @@ def operator_dispatch(
+                out_dt._spec = cast(DTensorSpec, output_specs[spec_idx])
+                out_dts.append(out_dt)
+                spec_idx += 1
         return tuple(out_dts) if len(out_dts) > 1 else out_dts[0]
```
General comment: This `return` assumes `len(out_dts) >= 1` or else there will be an index out of bounds error when accessing `out_dts[0]`. I am guessing that this `len(out_dts) >= 1` invariant always holds, but I just wanted to point this out. Maybe you want an assert?

wanchaol replied:
I'll land this as is as I don't want to trigger another full CI run, the assert is added in the follow up PR https://github.com/pytorch/pytorch/pull/90241

wanchaol replied:
nvm, might still need to change sth, will update.

awgu reviewed (commented):
I am not sure I have the context to approve, so I will defer to someone else.

awgu reviewed (approved):
This looks good to me, so I will stamp to unblock. Feel free to wait for others' review.

fduwjj reviewed (commented):
Overall this looks good to me and do we have an estimate of potential perf win?

wanchaol committed:
Update on "[dtensor] remove torchgen function schema and parse manually"


This PR get rids of torchgen FunctionSchema parsing and parse
it manually, it should resolve torchgen package issue and also
provide some perf wins when running DTensor eagerly

[ghstack-poisoned]

wanchaol commented:
> Overall this looks good to me and do we have an estimate of potential perf win?

I don't yet as I didn't profile this, but my suspection is that it could result in around 20-30 perf wins on DTensor execution, feel free to try it out, I'll focus on making other parts be faster if possible.

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -f "failure not related"
